core/src/main/java/io/questdb/cairo/TableSyncModel.java

```diff
@@ -0,0 +1,453 @@
+    }
+
+    public int getTableAction() {
+        return tableAction;
+    }
+
+    public void setTableAction(int tableAction) {
+        this.tableAction = tableAction;
+    }
+
+    public void setDataVersion(long dataVersion) {
+        this.dataVersion = dataVersion;
+    }
+
+    public void setMaxTimestamp(long maxTimestamp) {
+        this.maxTimestamp = maxTimestamp;
+    }
+
+    public void toBinary(TableWriterTask sink) {
+        sink.put(tableAction);
+        sink.put(dataVersion);
+        sink.put(maxTimestamp);
+
+        int n = columnTops.size();
+        sink.put(n); // column top count
+        if (n > 0) {
+            for (int i = 0; i < n; i += SLOTS_PER_COLUMN_TOP) {
+                sink.put(columnTops.getQuick(i)); // partition timestamp
+                sink.put((int) columnTops.getQuick(i + 1)); // column index
+                sink.put(columnTops.getQuick(i + 2)); // column top
+            }
+        }
+
+        n = varColumnSizes.size();
+        sink.put(n);
+        if (n > 0) {
+            for (int i = 0; i < n; i += SLOTS_PER_VAR_COLUMN_SIZE) {
+                sink.put(varColumnSizes.getQuick(i)); // partition timestamp
+                sink.put((int) varColumnSizes.getQuick(i + 1)); // column index
+                sink.put(varColumnSizes.getQuick(i + 2)); // column top
+            }
+        }
+
+        n = partitions.size();
+        sink.put(n); // partition count
+        for (int i = 0; i < n; i += SLOTS_PER_PARTITION) {
+            sink.put((int) partitions.getQuick(i)); // action
+            sink.put(partitions.getQuick(i + 1)); // partition timestamp
+            sink.put(partitions.getQuick(i + 2)); // start row
+            sink.put(partitions.getQuick(i + 3)); // row count
+            sink.put(partitions.getQuick(i + 4)); // name txn (suffix for partition name)
+            sink.put(partitions.getQuick(i + 5)); // data txn
+        }
+
+        n = addedColumnMetadata.size();
+        sink.put(n); // column metadata count - this is metadata only for column that have been added
+        for (int i = 0; i < n; i++) {
+            final TableColumnMetadata metadata = addedColumnMetadata.getQuick(i);
+            sink.put(metadata.getName());
+            sink.put(metadata.getType()); // column type
+            sink.put(metadata.getHash());
+            sink.put((byte) (metadata.isIndexed() ? 0 : 1)); // column indexed flag
+            sink.put(metadata.getIndexValueBlockCapacity());
+        }
+
+        n = columnMetaIndex.size();
+        sink.put(n); // column metadata shuffle index - drives rearrangement of existing columns on the slave
+        for (int i = 0; i < n; i += SLOTS_PER_COLUMN_META_INDEX) {
+            sink.put((int) columnMetaIndex.getQuick(i)); // action
+            sink.put(columnMetaIndex.getQuick(i + 1)); // (int,int) pair on where (from,to) column needs to move
+        }
+    }
+
+    @Override
+    public void toSink(CharSink sink) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 109 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks such as on lines 293-303, lines 315-334, lines 336-354, lines 356-379, lines 381-407 into separate methods.